src/plugins/discover/public/application/apps/main/utils/fetch_chart.ts

```diff
@@ -0,0 +1,121 @@
+import { ReduxLikeStateContainer } from '../../../../../../kibana_utils/common';
+import { sendErrorMsg, sendLoadingMsg } from '../services/use_saved_search_messages';
+
+export function fetchChart(
+  data$: SavedSearchData,
+  searchSource: SearchSource,
+  {
+    abortController,
+    appStateContainer,
+    data,
+    inspectorAdapters,
+    onResults,
+    searchSessionId,
+  }: {
+    abortController: AbortController;
+    appStateContainer: ReduxLikeStateContainer<AppState>;
+    data: DataPublicPluginStart;
+    inspectorAdapters: Adapters;
+    onResults: (foundDocuments: boolean) => void;
+    searchSessionId: string;
+  }
+) {
+  const { charts$, totalHits$ } = data$;
+
+  const interval = appStateContainer.getState().interval ?? 'auto';
+  const chartAggConfigs = updateSearchSource(searchSource, interval, data);
+
+  sendLoadingMsg(charts$);
+  sendLoadingMsg(totalHits$);
+
+  const fetch$ = searchSource
+    .fetch$({
+      abortSignal: abortController.signal,
+      sessionId: searchSessionId,
+      inspector: {
+        adapter: inspectorAdapters.requests,
+        title: i18n.translate('discover.inspectorRequestDataTitleChart', {
+          defaultMessage: 'Chart data',
+        }),
+        description: i18n.translate('discover.inspectorRequestDescriptionChart', {
+          defaultMessage:
+            'This request queries Elasticsearch to fetch the aggregation data for the chart.',
+        }),
+      },
+    })
+    .pipe(filter((res) => isCompleteResponse(res)));
+
+  fetch$.subscribe(
+    (res) => {
+      try {
+        const totalHitsNr = res.rawResponse.hits.total as number;
+        totalHits$.next({ fetchStatus: FetchStatus.COMPLETE, result: totalHitsNr });
+        onResults(totalHitsNr > 0);
+
+        const bucketAggConfig = chartAggConfigs.aggs[1];
+        const tabifiedData = tabifyAggResponse(chartAggConfigs, res.rawResponse);
+        const dimensions = getDimensions(chartAggConfigs, data);
+        const bucketInterval = search.aggs.isDateHistogramBucketAggConfig(bucketAggConfig)
+          ? bucketAggConfig?.buckets?.getInterval()
+          : undefined;
+        const chartData = buildPointSeriesData(tabifiedData, dimensions!);
+        charts$.next({
+          fetchStatus: FetchStatus.COMPLETE,
+          chartData,
+          bucketInterval,
+        });
+      } catch (e) {
+        charts$.next({
+          fetchStatus: FetchStatus.ERROR,
+          error: e,
+        });
+      }
+    },
+    (error) => {
+      if (error instanceof Error && error.name === 'AbortError') {
+        return;
+      }
+      sendErrorMsg(charts$, error);
+      sendErrorMsg(totalHits$, error);
+    }
+  );
+  return fetch$;
+}
+
+export function updateSearchSource(
+  searchSource: SearchSource,
+  interval: string,
+  data: DataPublicPluginStart
+) {
+  const indexPattern = searchSource.getField('index')!;
+  searchSource.setField('filter', data.query.timefilter.timefilter.createFilter(indexPattern!));
```

Comment:
:art: Nitpick:

The non null assertion is unnecessary, since we already have it in the line above.

Reply:
true, thx